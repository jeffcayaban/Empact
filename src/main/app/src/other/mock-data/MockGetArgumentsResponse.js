export default {
    "content": [
        {
            "id": 1,
            "isSupporting": true,
            "rootPetitionId": 1,
            "parentArgumentId": null,
            "agreesWithCQ": true,
            "criticalQuestionId": "CQ2",
            "creationDateTime": "2019-02-14T13:04:38Z",
            "lastUpdatedDateTime": "2019-02-14T13:04:51Z",
            "createdBy": {
                "id": 1,
                "username": "admin",
                "firstName": "Admin",
                "lastName": "Account",
                "authorities": [
                    "ROLE_ADMIN"
                ]
            },
            "argumentType": "PopularOpinion",
            "isAnonymous": false,
            "sources": [],
            "explanation": "TestExplanation",
        }
    ],
    "page": 0,
    "size": 5,
    "totalElements": 1,
    "totalPages": 1,
    "last": true
}