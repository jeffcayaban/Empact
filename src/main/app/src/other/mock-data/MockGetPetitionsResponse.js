export default {
    "content": [
        {
            "id": 1,
            "title": "TestTitle",
            "situation": "TestSituation",
            "action": "TestAction",
            "goal": "TestGoal",
            "value": "TestValue",
            "creationDateTime": "2019-02-14T13:04:29Z",
            "lastUpdatedDateTime": "2019-02-14T13:04:38Z",
            "createdBy": {
                "id": 1,
                "username": "admin",
                "firstName": "Admin",
                "lastName": "Account",
                "authorities": [
                    "ROLE_ADMIN"
                ]
            },
            "closingDateTime": "2019-02-15T13:04:27Z",
            "isAnonymous": false,
            "isWinning": true
        }
    ],
    "page": 0,
    "size": 5,
    "totalElements": 1,
    "totalPages": 1,
    "last": true
}