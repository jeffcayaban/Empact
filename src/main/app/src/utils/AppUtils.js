import {API_BASE_URL, AUTHORIZATION_HEADER_FIELD} from '../constants';
import {ACCESS_TOKEN, AUTHORIZATION_TOKEN_TYPE, PETITION_LIST_SIZE} from "../constants";

/** Creates a request object and sends the request to the server. **/
const request = (options) => {
    const headers = new Headers({ 'Content-Type': 'application/json' });
    const accessToken = localStorage.getItem(ACCESS_TOKEN);

    if (accessToken !== undefined && accessToken !== null) {
        headers.append(AUTHORIZATION_HEADER_FIELD, `${AUTHORIZATION_TOKEN_TYPE} ${accessToken}`)
    }

    options = {...options, headers: headers};

    return fetch(options.url, options).then(response =>
            (response.json()).then(jsonResponse => (!response.ok) ? Promise.reject(jsonResponse) : jsonResponse)
    );
};

// ---------------------------- APIs ----------------------------

/** Gets all the petitions **/
export function getAllPetitions(page, size, sort, showClosed) {
    page = page || 0;
    size = size || PETITION_LIST_SIZE;

    return request({
        url: API_BASE_URL + "/petitions?page=" + page + "&size=" + size + "&sort=" + sort + "&showClosed=" + showClosed,
        method: 'GET'
    });
}

/** Gets a single petition given an ID **/
export function getPetition(petitionId) {
    return request({
        url: API_BASE_URL + "/petitions/" + petitionId,
        method: 'GET'
    });
}

/** Creates a new petition given its details **/
export function createPetition(petitionData) {
    return request({
        url: API_BASE_URL + "/petitions/create",
        method: 'POST',
        body: JSON.stringify(petitionData)
    });
}

/** Deletes an existing petition given its details **/
export function deletePetition(deletePetitionData) {
    return request({
        url: API_BASE_URL + "/petitions/delete",
        method: 'DELETE',
        body: JSON.stringify(deletePetitionData)
    });
}

/** Updates an existing petition with the given details **/
export function editPetition(editPetitionData) {
    return request({
        url: API_BASE_URL + "/petitions/edit/",
        method: 'POST',
        body: JSON.stringify(editPetitionData)
    });
}

/** Logs in into a user account **/
export function login(loginRequest) {
    return request({
        url: API_BASE_URL + "/auth/signInUser",
        method: 'POST',
        body: JSON.stringify(loginRequest)
    });
}

/** Creates a new user account given the account details **/
export function signup(signupRequest) {
    return request({
        url: API_BASE_URL + "/auth/signUpUser",
        method: 'POST',
        body: JSON.stringify(signupRequest)
    });
}

/** Checks the username's availability given a username **/
export function checkUsernameAvailability(username) {
    return request({
        url: API_BASE_URL + "/user/checkUsernameAvailability?username=" + username,
        method: 'GET'
    });
}

/** Deletes a user **/
export function deleteUser(deleteUserRequest) {
    return request({
        url: API_BASE_URL + "/auth/deleteUser",
        method: 'DELETE',
        body: JSON.stringify(deleteUserRequest)
    })
}

/** Fetches the current user **/
export function getCurrentUser() {
    if(!localStorage.getItem(ACCESS_TOKEN)) {
        return Promise.reject("No access token set.");
    }

    return request({
        url: API_BASE_URL + "/user/getCurrentUser",
        method: 'GET'
    });
}

/** Fetches a user profile given a user's username **/
export function getUserProfile(username) {
    return request({
        url: API_BASE_URL + "/user/getUserProfile?username=" + username,
        method: 'GET'
    });
}

/** Fetches the petitions for a given user **/
export function getUserPetitions(page, size, sort, username, includeAnon, showClosed) {
    return request({
        url: API_BASE_URL + "/petitions/user?username=" + username + "&page=" + page + "&size=" + size + "&sort=" + sort
        + "&includeAnon=" + includeAnon + "&showClosed=" + showClosed,
        method: 'GET',
    });
}

/** Fetches the most discussed petitions for a given user **/
export function getMostDiscussedPetitionsByUser(page, size, sort, username, includeAnon, showClosed) {
    return request({
        url: API_BASE_URL + "/petitions/getMostDiscussedPetitionsByUser?username=" + username + "&page=" + page +
            "&size=" + size + "&sort=" + sort
            + "&includeAnon=" + includeAnon + "&showClosed=" + showClosed,
        method: 'GET',
    });
}

/** Closes a petition given its details **/
export function closePetition(closePetitionData) {
    return request({
        url: API_BASE_URL + "/petitions/close",
        method: 'POST',
        body: JSON.stringify(closePetitionData)
    })
}

/** Creates a new argument by expert opinion given the new details **/
export function createArgumentByExpertOpinion(newArgumentData) {
    return request({
        url: API_BASE_URL + "/arguments/createArgumentByExpertOpinion",
        method: 'POST',
        body: JSON.stringify(newArgumentData)
    });
}

/** Creates a new argument by popular opinion given the new details **/
export function createArgumentByPopularOpinion(newArgumentData) {
    return request({
        url: API_BASE_URL + "/arguments/createArgumentByPopularOpinion",
        method: 'POST',
        body: JSON.stringify(newArgumentData)
    });
}

/** Fetches the arguments for a given petition **/
export function getArgumentsByRootPetitionId(page, size, sort, petitionId, isSupporting) {
    return request({
        url: API_BASE_URL + "/arguments/getArgumentsByRootPetitionId?rootPetitionId=" + petitionId + "&isSupporting=" +
            isSupporting + "&page=" + page + "&size=" + size + "&sort=" + sort,
        method: 'GET',
    });
}

/** Fetches the arguments for a given argument **/
export function getArgumentsByArgumentId(page, size, sort, argumentId, isSupporting) {
    return request({
        url: API_BASE_URL + "/arguments/getArgumentsByArgumentId?argumentId=" + argumentId + "&isSupporting=" +
            isSupporting + "&page=" + page + "&size=" + size + "&sort=" + sort,
        method: 'GET',
    });
}

/** Fetches the argument for a given ID **/
export function getArgumentById(argumentId) {
    return request({
        url: API_BASE_URL + "/arguments/" + argumentId,
        method: 'GET'
    });
}

/** Fetches the parent argument for a given argument **/
export function getParentArgumentById(argumentId) {
    return request({
        url: API_BASE_URL + "/arguments/getParentArgumentById?argumentId=" + argumentId,
        method: 'GET',
    });
}

/** Checks whether a petition is closed or not **/
export function isPetitionClosed(petitionId) {
    return request({
        url: API_BASE_URL + "/petitions/isPetitionClosed/" + petitionId,
        method: 'GET'
    })
}

/** Fetches the arguments for a given user **/
export function getUserArguments(page, size, sort, username, showAnonymousContent) {
    return request({
        url: API_BASE_URL + "/arguments/getArgumentsByUsernameAndAnonymity?username=" + username +
            "&showAnonymousContent=" + showAnonymousContent + "&page=" + page + "&size=" + size + "&sort=" + sort,
        method: 'GET'
    })
}

/** Fetches the number of supporting or opposing arguments for a given petition **/
export function countArgsByPetitionIdAndIsSupporting(petitionId, isSupporting) {
    return request({
        url: API_BASE_URL + "/arguments/countArgsByPetitionIdAndIsSupporting?petitionId=" + petitionId
            + "&isSupporting=" + isSupporting,
        method: 'GET',
    })
}

/** Fetches the number of supporting or opposing arguments for a given argument **/
export function countArgsByArgumentIdAndIsSupporting(argumentId, isSupporting) {
    return request({
        url: API_BASE_URL + "/arguments/countAllByArgumentIdAndIsSupporting?argumentId=" + argumentId
            + "&isSupporting=" + isSupporting,
        method: 'GET',
    })
}

/** Deletes an argument given its details **/
export function deleteArgument(deleteArgumentData) {
    return request({
        url: API_BASE_URL + "/arguments/delete",
        method: 'DELETE',
        body: JSON.stringify(deleteArgumentData)
    });
}

/** Updates an existing argument by popular opinion with new details **/
export function editArgumentByPopularOpinion(editArgumentData) {
    return request({
        url: API_BASE_URL + "/arguments/editArgumentByPopularOpinion",
        method: 'POST',
        body: JSON.stringify(editArgumentData)
    });
}

/** Updates an existing argument by expert opinion with new details **/
export function editArgumentByExpertOpinion(editArgumentData) {
    return request({
        url: API_BASE_URL + "/arguments/editArgumentByExpertOpinion",
        method: 'POST',
        body: JSON.stringify(editArgumentData)
    });
}

/** Fetches the most discussed petitions **/
export function getMostDiscussedPetitions(page, size, sort, showClosed) {
    return request({
        url: API_BASE_URL + "/petitions/getMostDiscussedPetitions?page=" + page + "&size=" + size + "&sort=" + sort +
            "&showClosed=" + showClosed,
        method: 'GET'
    });
}

/** Fetches the most discussed arguments **/
export function getMostDiscussedArguments(pageSize, showClosed) {
    return request({
        url: API_BASE_URL + "/arguments/getMostDiscussedArguments?size=" + pageSize + "&showClosed=" + showClosed,
        method: 'GET'
    });
}

/** Fetches the most discussed arguments for a given petition **/
export function getMostDiscussedArgumentsForPetition(page, size, sort, petitionId, isSupporting) {
    return request({
        url: API_BASE_URL + "/arguments/getMostDiscussedArgsForPetition?page=" + page + "&size=" + size + "&sort=" + sort +
            "&petitionId=" + petitionId + "&isSupporting=" + isSupporting,
        method: 'GET'
    });
}

/** Fetches the most discussed arguments for a given argument **/
export function getMostDiscussedArgumentsForArgument(page, size, sort, argumentId, isSupporting) {
    return request({
        url: API_BASE_URL + "/arguments/getMostDiscussedArgsForArgument?page=" + page + "&size=" + size + "&sort=" + sort +
            "&argumentId=" + argumentId + "&isSupporting=" + isSupporting,
        method: 'GET'
    });
}

/** Fetches the most discussed arguments for a given user **/
export function getMostDiscussedArgumentsForUser(page, size, sort, username, showAnonymousContent) {
    return request({
        url: API_BASE_URL + "/arguments/getMostDiscussedArgsForUser?page=" + page + "&size=" + size + "&sort=" + sort +
            "&username=" + username + "&showAnonymousContent=" + showAnonymousContent,
        method: 'GET'
    });
}

/** Fetches the data nodes required for the data visualisation of arguments **/
export function getDataNodes(contentId, isPetition) {
    return request({
        url: API_BASE_URL + "/arguments/getDataNodes?contentId=" + contentId + "&isPetition=" + isPetition,
        method: 'GET'
    });
}

/** Fetches data required for the petition report **/
export function getPetitionReport(petitionId) {
    return request({
        url: API_BASE_URL + "/arguments/getPetitionReport?petitionId=" + petitionId,
        method: 'GET'
    });
}

/** Searches for petitions given a query **/
export function searchPetitions(page, size, sort, showClosed, query) {
    return request({
        url: API_BASE_URL + "/petitions/search?page=" + page + "&size=" + size + "&sort=" + sort + "&showClosed=" + showClosed
         + "&query=" + query,
        method: 'GET'
    });
}
