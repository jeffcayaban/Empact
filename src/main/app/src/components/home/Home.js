import React from 'react'

//  Components
import ArgumentEntry from "../common/ArgumentEntry";
import HomeHeader from "./HomeHeader";
import Loader from "../Loader";
import PetitionEntry from "../common/PetitionEntry";
import TopContent from "./TopContent";

// Other
import { getMostDiscussedArguments, getMostDiscussedPetitions } from "../../utils/AppUtils";
import {
    ARGUMENT,
    DESCENDING,
    NO_OF_MOST_DISCUSSED_ARGUMENTS,
    NO_OF_MOST_DISCUSSED_PETITIONS,
    PETITION
} from "../../constants";

/**
 * Represents the home page of the system.
 */

class Home extends React.PureComponent {

    constructor(props) {
        super(props);

        this.state = {
            isLoading: true,
            mostDiscussedPetitionsSize: NO_OF_MOST_DISCUSSED_PETITIONS,
            mostDiscussedPetitions: [],
            mostDiscussedArgumentsSize: NO_OF_MOST_DISCUSSED_ARGUMENTS,
            mostDiscussedArguments: [],
        }
    }

    // Fetches the most discussed petitions.
    getMostDiscussedPetitions() {
        return getMostDiscussedPetitions(0, this.state.mostDiscussedPetitionsSize, DESCENDING, false)
    }

    // Fetches the most discussed arguments.
    getMostDiscussedArguments() {
        return getMostDiscussedArguments(this.state.mostDiscussedArgumentsSize, false)
    }

    // Generates an individual argument entry for an arguments table.
    static generateArgumentEntry(argument) {
        return (
            <ArgumentEntry showCreatorName={true} showCreatorBadge={false} key={argument.id} argument={argument}
                isSupporting={argument.isSupporting} />
        )
    }

    // Generates an individual petition entry for a petitions table.
    static generatePetitionEntry(petition) {
        return <PetitionEntry showCreatorName={true} key={petition.id} petition={petition} showCreationDetails={true} />
    }

    componentWillMount() {
        // Fetch the most discussed petitions and arguments.
        const fetches = [ this.getMostDiscussedPetitions(), this.getMostDiscussedArguments() ];

        Promise.all(fetches)
            .then((data) => {
                this.setState({
                    isLoading: false,
                    mostDiscussedPetitions: data[0].content,
                    mostDiscussedArguments: data[1].content
                })
            })
            .catch(() => this.setState({ isLoading: false }))
    }

    render() {
        if (!this.state.isLoading) {
            // Assumes that data fetching has completed.
            const { mostDiscussedPetitions, mostDiscussedArguments } = this.state;

            return (
                <div>
                    <HomeHeader />
                    <div>
                        <TopContent contentType={PETITION}
                            mostDiscussedContent={mostDiscussedPetitions}
                            contentMapFn={Home.generatePetitionEntry}
                        />
                        <TopContent
                            contentType={ARGUMENT}
                            mostDiscussedContent={mostDiscussedArguments}
                            contentMapFn={Home.generateArgumentEntry}
                        />
                    </div>
                </div>
            );
        } else {
            // Assumes that data is being fetched.
            return <Loader isLoading={this.state.isLoading} />
        }
    }
}

export default Home;