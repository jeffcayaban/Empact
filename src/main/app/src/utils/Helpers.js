import React from 'react'

import moment from "moment";
import swal from "sweetalert2";
import {
    ARGUMENTATION_SCHEMES,
    CLOSE,
    CRITICAL_QUESTION_IDS, CURRENT_USERNAME,
    DELETE,
    EXPERT_OPINION,
    POPULAR_OPINION,
    ROLE_ADMIN,
    VALID_URL_PREFIX_REGEX
} from "../constants";
import {ACCESS_TOKEN} from "../constants";
import PopularOpinion from "../components/argument/create-argument/forms/PopularOpinion";
import ExpertOpinion from "../components/argument/create-argument/forms/ExpertOpinion";
import CreatedByOn from "../components/common/CreatedByOn";
import AcceptedCQ
    from "../components/argument/create-argument/create-argument-for-argument/popular-opinion-cqs/AcceptedCQ";
import ExpertiseCQ
    from "../components/argument/create-argument/create-argument-for-argument/expert-opinion-cqs/ExpertiseCQ";
import {login} from "./AppUtils";

/** Checks whether the user is logged in. **/
export function isUserLoggedIn(isAuthenticated) {
    return isAuthenticated && localStorage.getItem(CURRENT_USERNAME) !== null;
}

/** Checks whether the user is an administrator **/
export function isUserAdmin(currentUser, isAuthenticated) {
    return isUserLoggedIn(isAuthenticated) && currentUser.authorities[0] === ROLE_ADMIN;
}

/** Formats a timestamp in DD-MM-YYYY **/
export function formatTimestampToDate(timestamp) {
    return moment(timestamp).format('DD-MM-YYYY');
}

/** Formats a timestamp in DD-MM-YYYY h:mma **/
export function formatTimestampToDateTime(timestamp) {
    return moment(timestamp).format('DD-MM-YYYY h:mma');
}

/** Generates a set of labels to be displayed for an argument. **/
export function getArgumentLabels(argument, parentArgument, agreesWithCQ) {
    const criticalQuestionId = argument.criticalQuestionId;
    const argumentType = argument.argumentType;
    const toReturn = [];

    if (criticalQuestionId === CRITICAL_QUESTION_IDS[0]) {
        // Critical Question: Are the circumstances true?

        const isSupportingText = agreesWithCQ ? "true" : "false";
        toReturn[0] = `The situation is ${isSupportingText}`;

    } else if (criticalQuestionId === CRITICAL_QUESTION_IDS[1]) {
        // Critical Question: Does the action achieve the goal?

        const isSupportingText = agreesWithCQ ? "" : " not";
        toReturn[0] = `The action does${isSupportingText} achieve the goal`;
    }

    if (argument.subCriticalQuestionId && parentArgument) {
        // The argument is made to another argument.

        const subCriticalQuestionId = argument.subCriticalQuestionId;

        if (subCriticalQuestionId === CRITICAL_QUESTION_IDS[2]) {
            // Critical Question: Is the expert credible?

            const isSupportingText = agreesWithCQ ? "" : " not";
            toReturn[2] = "The expert, " + parentArgument.expert + `, is${isSupportingText} credible`;
        } else if (subCriticalQuestionId === CRITICAL_QUESTION_IDS[3]) {
            // Critical Question: Is the explanation generally accepted?

            const isSupportingText = agreesWithCQ ? "" : " not";
            toReturn[2] = `The explanation is${isSupportingText} accepted`;
        }
    }

    if (argumentType === EXPERT_OPINION) {
        toReturn[1] = `Argument by Expert Opinion`;
    } else if (argumentType === POPULAR_OPINION) {
        toReturn[1] = 'Argument by Popular Opinion';
    }

    return toReturn;
}

/** Pluralises a time unit. **/
function pluraliseTime(time, label) {
    return (time === 1) ? label : label + 's';
}

/** Generates a readable date message **/
export function getTimeRemaining(timestamp) {
    const dateToCompare = moment(timestamp);
    const now = moment();

    const exaggerateTimeStyle = { color: 'red', marginBottom: '0px' };

    if (dateToCompare.diff(now, 'seconds') < 60) {
        // Focuses on Seconds.
        const time = dateToCompare.diff(now, 'seconds');
        return <p style={exaggerateTimeStyle}>{`${time} ${pluraliseTime(time, 'second')}`}</p>;
    } else if (dateToCompare.diff(now, 'minutes') < 60) {
        // Focuses on Minutes.
        const time = dateToCompare.diff(now, 'minutes');
        return <p style={exaggerateTimeStyle}>{`${time} ${pluraliseTime(time, 'minute')}`}</p>;
    } else if (dateToCompare.diff(now, 'hours') < 24) {
        // Focuses on Hours.
        const time = dateToCompare.diff(now, 'hours');
        return <p style={exaggerateTimeStyle}>{`${time} ${pluraliseTime(time, 'hour')}`}</p>;
    } else if (dateToCompare.diff(now, 'days') < 31) {
        // Focuses on Days.
        const time = dateToCompare.diff(now, 'days');
        return `${time} ${pluraliseTime(time, 'day')}`
    } else if (dateToCompare.diff(now, 'weeks') < 4) {
        // Focuses on Weeks.
        const time = dateToCompare.diff(now, 'weeks');
        return `${time} ${pluraliseTime(time, 'week')}`
    } else if (dateToCompare.diff(now, 'months') < 12) {
        // Focuses on Months.
        const time = dateToCompare.diff(now, 'months');
        return `${time} ${pluraliseTime(time, 'month')}`
    } else {
        // Focuses on Years.
        const time = dateToCompare.diff(now, 'years');
        return `${time} ${pluraliseTime(time, 'year')}`
    }
}

/** Determines whether the date given has passed. **/
export function isDatePassed(timestamp) {
    return moment(timestamp).diff(moment()) <= 0;
}

/** Generates a successful dialog for argument creation. **/
export function showArgCreationConfirmation(response) {
    return (
        swal({
            title: 'Success!',
            text: 'Your argument has been created!',
            type: 'success',
            confirmButtonText: 'Your Argument',
            cancelButtonText: 'Close',
            showCancelButton: true,
        }).then((result) => {
            if (result.value) {
                window.location = `/argument/${response.argumentId}`
            } else {
                window.location.reload();
            }
        })
    )
}

/** Generates a successful dialog for argument creation. **/
export function renderArgumentCreatorName(argument) {
    return (!argument.isAnonymous) ? argument.createdBy.firstName : "The argument creator";
}

/** Generates a dialog to confirm the user's action to either delete or close some content. **/
export function generateConfirmActionDialog(contentType, contentAction) {
    let displayedAction;

    if (contentAction === DELETE) {
        // Assumes the user's action to be regarding deletion.
        displayedAction = " deleting";

    } else if (contentAction === CLOSE) {
        // Assumes the user's action to be regarding closing a petition.
        displayedAction = " closing"
    }

    return (
        swal({
            title: 'Are you sure?',
            text: `Do you wish to proceed with${displayedAction} this ${contentType.toLowerCase()}?`,
            type: 'warning',
            showCancelButton: true,
            confirmButtonColor: '#3085d6',
            cancelButtonColor: '#d33',
            confirmButtonText: 'Yes'
        })
    );
}

/** Generates a dialog for when something went wrong. **/
export function generateSomethingWentWrongDialog() {
    return generateErrorDialog('Sorry! Something went wrong. Please try again');
}

/** Generates a dialog for when something went wrong. **/
export function generateAdminActionSuccessDialog(contentType, contentAction) {
    let displayedAction;

    if (contentAction === DELETE) {
        displayedAction = "deleted";
    } else if (contentAction === CLOSE) {
        displayedAction = "closed"
    }

    return (
        swal({ title: 'Success!', text: `The ${contentType.toLowerCase()} has been ${displayedAction}.`, type: 'success' })
            .then(() => window.location.reload())
    )
}

/** Generates a dialog to confirm the user's action. **/
export function generateGeneralConfirmDialog(message = "You won't be able to revert this!") {
    return (
        swal({
            title: 'Are you sure?',
            text: message,
            type: 'warning',
            showCancelButton: true,
            confirmButtonColor: '#3085d6',
            cancelButtonColor: '#d33',
            confirmButtonText: 'Yes'
        })
    )
}

/** Generates a generic error dialog. **/
export function generateErrorDialog(message) {
    return swal('Error', message, 'error');
}

/** Generates a generic success dialog. **/
export function generateSuccessDialog(message, confirmButtonText) {
    return swal({ title: 'Success!', text: message, type: 'success', confirmButtonText: confirmButtonText });
}

/** Generates the appropriate critical question for an argument **/
export function generateCriticalQuestions(argument, userSupportsArgument) {
    const argumentType = argument.argumentType;

    if (argumentType === EXPERT_OPINION) {
        return <div><ExpertiseCQ argument={argument} isSupporting={userSupportsArgument} /></div>;
    } else if (argumentType === POPULAR_OPINION) {
        return <div><AcceptedCQ argument={argument} isSupporting={userSupportsArgument} /></div>;
    }
}

/** Checks whether a URL source is of the correct format. E.g. Starting with http or https **/
export function checkSourceUrlsIsValid(sources) {
    if (sources !== undefined && sources !== null) {
        let i;
        for (i = 0; i < sources.length; i++) {
            if (!VALID_URL_PREFIX_REGEX.test(sources[i])) {
                return false;
            }
        }
    }
    return true;
}

/** Checks whether a URL source is of the correct format. E.g. Starting with http or https **/
export function handleArgumentSubmit(sources, isForEdit, editArgRequest, editArgApi, createArgRequest, createArgApi,
                                     argumentId, argument) {

    if (checkSourceUrlsIsValid(sources)) {
        // Assumes that the URL sources for the argument are in the correct format.

        swal({
            title: 'Are you sure?',
            text: 'Do you wish to proceed with the details of your argument?',
            type: 'warning',
            showCancelButton: true,
            confirmButtonColor: '#3085d6',
            cancelButtonColor: '#d33',
            confirmButtonText: 'Yes'
        }).then((result) => {
            if (result.value) {
                if (isForEdit) {
                    // For argument modification.
                    let existNoChanges;

                    if (argument.argumentType === EXPERT_OPINION) {
                        existNoChanges = compareExpertOpinionData(argument, editArgRequest);
                    } else {
                        existNoChanges = comparePopularOpinionData(argument, editArgRequest);
                    }

                    if (!existNoChanges) {
                        editArgApi(editArgRequest)
                            .then(response => {
                                swal({
                                    title: 'Success!',
                                    text: 'Your argument has been updated!',
                                    type: 'success',
                                    confirmButtonText: 'Your Argument',
                                    cancelButtonText: 'Close',
                                    showCancelButton: true,
                                }).then((result) => {
                                    if (result.value) {
                                        window.location = `/argument/${argumentId}`
                                    } else {
                                        window.location.reload();
                                    }
                                })
                            })
                            .catch(() => generateErrorDialog('Your argument couldn\'t be updated.'))
                    } else {
                        generateErrorDialog("There are no new changes to be applied.")
                    }


                } else {
                    // For argument creation.
                    createArgApi(createArgRequest)
                        .then(response => showArgCreationConfirmation(response))
                        .catch(() => generateErrorDialog('Your argument couldn\'t be created.'))
                }
            }
        });
    } else {
        // Assumes that the URL sources for the argument are not in the correct format, so an error is displayed.
        generateErrorDialog('Your argument couldn\'t be created. Your sources contain an invalid URL.');
    }
}

/** Displays the appropriate argumentation scheme form. **/
export function showArgumentTemplate(statement, selectedArgScheme, petitionId, isSupporting, userAgreeWithQuestion,
                                     argumentId, criticalQuestion, subCriticalQuestion) {
    if (selectedArgScheme) {
        let argumentTemplate;

        switch(selectedArgScheme) {
            case ARGUMENTATION_SCHEMES[0]:
                argumentTemplate = <ExpertOpinion
                    argumentId={argumentId}
                    petitionId={petitionId}
                    isSupporting={isSupporting}
                    criticalQuestionId={criticalQuestion}
                    agreesWithCQ={userAgreeWithQuestion}
                    subCriticalQuestionId={subCriticalQuestion}
                />;
                break;
            case ARGUMENTATION_SCHEMES[1]:
                argumentTemplate = <PopularOpinion
                    argumentId={argumentId}
                    petitionId={petitionId}
                    isSupporting={isSupporting}
                    criticalQuestionId={criticalQuestion}
                    agreesWithCQ={userAgreeWithQuestion}
                    statement={statement}
                    subCriticalQuestionId={subCriticalQuestion}
                />;
                break;
            default:
                argumentTemplate = null;
                break;
        }

        return (
            <div style={{ paddingTop: '30px' }}>
                {argumentTemplate}
            </div>
        );
    }
}

/** Check whether any changes have been made between two sets of petition data. **/
export function comparePetitionData(existingData, newData) {
    const actionIsSame = existingData.action === newData.action;
    const goalIsSame = existingData.goal === newData.goal;
    const situationIsSame = existingData.situation === newData.situation;
    const valueIsSame = existingData.value === newData.value;
    const titleIsSame = existingData.title === newData.title;
    const isAnonymousSame = existingData.isAnonymous === newData.isAnonymous;
    const isClosingDateSame = moment(existingData.closingDateTime).valueOf() === newData.closingDateTime;

    return actionIsSame && goalIsSame && situationIsSame && valueIsSame && titleIsSame && isAnonymousSame && isClosingDateSame;
}

/** Check whether any changes have been made between two sets of "argument by expert opinion" data. **/
export function compareExpertOpinionData(existingData, newData) {
    const expertIsSame = existingData.expert === newData.expert;
    const expertDomainIsSame = existingData.expertDomain === newData.expertDomain;
    const expertAssertionIsSame = existingData.expertAssertion === newData.expertAssertion;
    const sourcesIsSame = JSON.stringify(existingData.sources) === JSON.stringify(newData.sources);
    const isAnonymousSame = existingData.isAnonymous === newData.isAnonymous;

    return expertIsSame && expertDomainIsSame && expertAssertionIsSame && sourcesIsSame && isAnonymousSame;

}

/** Check whether any changes have been made between two sets of "argument by popular opinion" data. **/
export function comparePopularOpinionData(existingData, newData) {
    const explanationIsSame = existingData.explanation === newData.explanation;
    const sourcesIsSame = JSON.stringify(existingData.sources) === JSON.stringify(newData.sources);
    const isAnonymousSame = existingData.isAnonymous === newData.isAnonymous;

    return explanationIsSame && sourcesIsSame && isAnonymousSame;
}

/** Truncates a piece of text by a given length. **/
export function truncateText(text, length) {
    return (length && text) && (text.length > length ? `${text.substr(0, length)}...` : text);
}

/** Determines whether an argument agrees with the critical question. **/
export function getIsArgumentAgreeWithCQ(isSupporting, ArgAgreesWithCQ) {
    return (isSupporting) ? ArgAgreesWithCQ : !ArgAgreesWithCQ;
}

/** Generates the argument creator's name **/
export function generateArgumentCreatorName(argument, showCreatorName) {
    const creationDate = formatTimestampToDate(argument.creationDateTime);

    if (showCreatorName) {
        return <CreatedByOn content={argument} date={creationDate} />;
    } else {
        return <span>{`Created on ${creationDate}`}</span>;
    }
}

/** Sends the request to login into a given account. **/
export function sendLoginRequest(request, goToHome, username, loginCallback) {
    login(request)
        .then(response => {
            localStorage.setItem(ACCESS_TOKEN, response.accessToken);
            localStorage.setItem(CURRENT_USERNAME, username);
            loginCallback !== null ? loginCallback() : window.location = `/`
        }).catch(error => {

        if (error.status === 401) {
            generateErrorDialog('Your Username or Password is incorrect. Please try again');
        } else {
            generateSomethingWentWrongDialog();
        }

        if (goToHome) {
            // Redirects the user to the home page.
            window.location = `/`;
        }
    });
}
