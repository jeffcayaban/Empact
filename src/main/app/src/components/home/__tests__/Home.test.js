import React from 'react';
import toJson from "enzyme-to-json";
import { shallow } from "enzyme";
import Home from "../Home";

import * as AppUtils from '../../../utils/AppUtils';

import MockGetArgumentsResponse from '../../../other/mock-data/MockGetArgumentsResponse';
import MockGetPetitionsResponse from '../../../other/mock-data/MockGetPetitionsResponse';

describe('Home', () => {

    it('should render correctly', () => {
        const originalGetMostDiscussedPetitions = AppUtils.getMostDiscussedPetitions;
        const originalGetMostDiscussedArguments = AppUtils.getMostDiscussedArguments;

        AppUtils.getMostDiscussedPetitions = () => MockGetPetitionsResponse;
        AppUtils.getMostDiscussedArguments = () => MockGetArgumentsResponse;

        const component = shallow(<Home />);
        component.setState({ isLoading: false });

        expect(toJson(component)).toMatchSnapshot();

        AppUtils.getMostDiscussedPetitions = originalGetMostDiscussedPetitions;
        AppUtils.getMostDiscussedArguments = originalGetMostDiscussedArguments;
    });

    it('should render the loader if isLoading is true', () => {
        const component = shallow(<Home />);
        component.setState({ isLoading: true });
        expect(toJson(component)).toMatchSnapshot();
    });
});
