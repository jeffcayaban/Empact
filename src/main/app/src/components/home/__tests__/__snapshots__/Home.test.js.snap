// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should render correctly 1`] = `
<div>
  <HomeHeader />
  <div>
    <TopContent
      contentMapFn={[Function]}
      contentType="Petition"
      mostDiscussedContent={Array []}
    />
    <TopContent
      contentMapFn={[Function]}
      contentType="Argument"
      mostDiscussedContent={Array []}
    />
  </div>
</div>
`;

exports[`Home should render the loader if isLoading is true 1`] = `
<Loader
  isLoading={true}
/>
`;
